WMF Framework/WMFArticle+Migration.swift

```diff
@@ -0,0 +1,299 @@
+                if let limit { request.fetchLimit = limit }
+
+                let articles = try wikipediaContext.fetch(request)
+                guard !articles.isEmpty else { return }
+
+                var snapshots: [SavedArticleSnapshot] = []
+                snapshots.reserveCapacity(articles.count)
+
+                for article in articles {
+                    guard
+                        let savedDate = article.savedDate,
+                        let ids = self.getIdFromLegacyArticleURL(article)
+                    else { continue }
+
+                    snapshots.append(
+                        SavedArticleSnapshot(
+                            ids: ids,
+                            savedDate: savedDate,
+                            viewedDate: article.viewedDate
+                        )
+                    )
+                }
+
+                guard !snapshots.isEmpty else { return }
+
+                var applyError: Error?
+                wmfContext.performAndWait {
+                    do {
+                        for snap in snapshots {
+                            try self.applySavedStateOnWMFContext(snapshot: snap, in: wmfContext, wmfDataStore: wmfDataStore)
+                        }
+                        if wmfContext.hasChanges { try wmfContext.save() }
+                    } catch {
+                        applyError = error
+                    }
+                }
+                if let applyError { throw applyError }
+
+                for article in articles {
+                    article.isSavedMigrated = true
+                }
+                if wikipediaContext.hasChanges { try wikipediaContext.save() }
+
+            } catch {
+                DDLogError("WMFArticle isSavedMigrated migration error: \(error)")
+            }
+        }
+    }
+
+    // MARK: - Update Saved State
+
+    @objc func unsave(forArticleObjectID objectID: NSManagedObjectID) {
+        guard let wmfDataStore = WMFDataEnvironment.current.coreDataStore else { return }
+
+        dataStore.performBackgroundCoreDataOperation { wikipediaContext in
+            wikipediaContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+
+            guard let wmfContext = try? wmfDataStore.newBackgroundContext else { return }
+            wmfContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+
+            do {
+                guard let article = try? wikipediaContext.existingObject(with: objectID) as? WMFArticle else {
+                    DDLogError("Revert failed: article not found for objectID \(objectID)")
+                    return
+                }
+
+                article.isSavedMigrated = false
+                article.savedDate = nil
+
+                var applyError: Error?
+                wmfContext.performAndWait {
+                    do {
+                        try self.syncSavedState(from: article, to: wmfContext, action: .unsave, wmfDataStore: wmfDataStore)
```

Comment:
I think you are still passing NSManagedObjects across contexts here. When you call article.viewedDate within `syncSavedState`, I think you are accessing it from the `wmfContext` here.